test/utils_test.js

```diff
@@ -242,6 +242,207 @@ describe("utils", () => {
         });
       });
     });
+
+    it("should use local default in favor of ref default for object", () => {
+      const schema = {
+        definitions: {
+          testdef: {
+            type: "object",
+            properties: {
+              foo: { type: "number" },
+            },
+            default: { foo: 1 },
+          },
+        },
+        $ref: "#/definitions/testdef",
+        default: { foo: 42 },
+      };
+
+      expect(getDefaultFormState(schema, undefined, schema.definitions)).eql({
+        foo: 42,
+      });
+    });
+
+    it("should work for nested array", () => {
+      const schema = {
+        type: "object",
+        properties: {
+          arr: {
+            type: "array",
+            items: {
+              type: "boolean",
+              default: false,
+            },
+          },
+        },
+      };
+
+      expect(
+        getDefaultFormState(schema, {
+          arr: [undefined],
+        })
+      ).eql({
+        arr: [false],
+      });
+    });
+
+    it("should work for arrays of objects with defaults", () => {
+      const schema = {
+        type: "object",
+        properties: {
+          arr: {
+            type: "array",
+            items: {
+              type: "object",
+              properties: {
+                defaultFilled: {
+                  type: "boolean",
+                  default: true,
+                },
+              },
+            },
+          },
+        },
+      };
+
+      expect(
+        getDefaultFormState(schema, {
+          arr: [{}],
+        })
+      ).eql({
+        arr: [{ defaultFilled: true }],
+      });
+    });
+
+    it("should use additionalItems defaults for items not specified in items", () => {
+      const schema = {
+        type: "object",
+        properties: {
+          arr: {
+            type: "array",
+            items: [
+              {
+                type: "boolean",
+                default: false,
+              },
+            ],
+            additionalItems: {
+              type: "string",
+              default: "foobar",
+            },
+          },
+        },
+      };
+
+      expect(
+        getDefaultFormState(schema, {
+          arr: [undefined, undefined],
+        })
+      ).eql({
+        arr: [false, "foobar"],
+      });
+    });
+
+    it("should use array default in favor of array item default", () => {
+      const schema = {
+        type: "array",
+        default: [1, 2],
+        items: {
+          type: "number",
+          default: 3,
+        },
+      };
+
+      expect(getDefaultFormState(schema, [0, undefined])).eql([0, 2]);
```

Comment:
Same concern here about filling in elements of `default` when the field itself already exists.